FD refactor + sub device support

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/13655

### Problem description
Various aspects of FD/runtime need to be updated to support multiple sub devices.
SubDevice: Container of the various cores we want to group together from a device
SubDeviceManager: Container of the state/objects that represent a configuration of the device that is split into SubDevices.

### What's changed
Support multiple dispatch entries for worker->dispatch sync

Update dispatch d/s to have a semaphore per dispatch entry to enable syncing on specific worker counts

Update LaunchMessageRingBufferState and WorkerConfigBufferMgr to be tracked per sub_device

Update various FD commands to support syncing on multiple sub devices:
- ERB, EWB, ERE, Synchronize/finish takes in a list of sub devices for blocking/issuing waits on. Will wait on all sub-devices if none are provided
- Trace will track only specific sub devices used
- EP currently only supports one sub-device

Update allocator to support taking in a CoreRangeSet for banks, instead of assuming a rectangular grid

Add support for splitting a device into multiple SubDevices, as well and maintaining different SubDeviceManager configurations, owned by device

Add basic tests to validate sub-device support

I will be continuing to add/updates tests in the near future, but would like to get this merged since this touches a lot of files/state to avoid dealing with conflicts.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11823639393
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

tt-aho committed:
#13655: Initial FD refactor to support sub devices
Support multiple dispatch entries for worker->dispatch sync
Update dispatch d/s to have a semaphore per dispatch entry to enable syncing on specific worker counts
Update LaunchMessageRingBufferState and WorkerConfigBufferMgr to be tracked per sub_device
Update various FD commands to support syncing on multiple sub devices:
- ERB, EWB, ERE takes in a list of sub devices for blocking/issuing waits on. Will wait on all sub-devices if none are provided
- Trace will track only specific sub devices used
- EP currently only supports one sub-device
- Remove compile time mcast grid and unicast cores from dispatch kernels
  CQDispatchGoSignalMcastCmd now expects noc txn data to follow the cmd for sending go signal to cores

pgkeller commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1581,11 +1600,22 @@ void EnqueueRecordEventCommand::process() {
+    uint32_t dispatch_message_base_addr = dispatch_constants::get(
         dispatch_core_type).get_device_command_queue_addr(CommandQueueDeviceAddrType::DISPATCH_MESSAGE);
 
+    uint32_t last_index = this->expected_num_workers_completed.size() - 1;
```
don't follow what this is doing.
code here will conflict w/ the kernel bin ring buffer async changes

tt-aho replied:
The code change here is that there are now multiple worker completion counters. Previously we only had one counter for the whole chip, and we would only need a single wait cmd. We now enqueue a wait cmd per counter (todo to take in a list of which counters to wait on), and we specialize the last iteration because we only need a write barrier on the dispatch core once at the end, and is not needed per wait cmd.

What would be the conflict with the ring buffer changes?

pgkeller reviewed (commented):
only got part way through, will review again later

tt-aho commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1581,11 +1600,22 @@ void EnqueueRecordEventCommand::process() {
         dispatch_core_type).get_device_command_queue_addr(CommandQueueDeviceAddrType::DISPATCH_MESSAGE);
 
+    uint32_t last_index = this->expected_num_workers_completed.size() - 1;
+    // We only need the write barrier for the last wait cmd
```
Possible. I didn't commonize it since they use different fns `add_dispatch_wait` vs `add_dispatch_wait_with_prefetch_stall` and could potentially take different args, and wanted to avoid adding a wrapper fn where we'd need to match default args with the actual fns to keep same functionality.

tt-aho commented on tt_metal/impl/device/device.cpp:
```diff
@@ -3102,6 +3126,29 @@ void Device::check_allocator_is_initialized() const {
     }
 }
 
+void Device::reset_num_sub_devices(uint32_t num_sub_devices) {
```
Yes. I tested locally by inserting it into the setup/teardown process for device and running a few tests to verify functionality, but nothing uses it in the actual pr.

aliuTT commented on tt_metal/impl/dispatch/command_queue_interface.hpp:
```diff
@@ -64,8 +65,12 @@ struct dispatch_constants {
+
     static constexpr uint8_t MAX_NUM_HW_CQS = 2;
-    typedef uint16_t prefetch_q_entry_type;
+    static constexpr uint32_t DISPATCH_MESSAGE_ENTRIES = 16;
```
Is this the max number of sub devices? It seems like it's used across the stack to represent that. It may be clearer to have another `MAX_NUM_SUBDEVICES = DISPATCH_MESSAGE_ENTRIES`.

cfjchu replied:
+1 I was looking at device changes and was wondering what `DISPATCH_MESSAGE_ENTRIES` was supposed to be.

tt-aho replied:
I've added a new static constexpr in Device, `static constexpr uint32_t MAX_NUM_SUB_DEVICES = dispatch_constants::DISPATCH_MESSAGE_ENTRIES;`, and updated the relevant use cases in device files to use this instead of `DISPATCH_MESSAGE_ENTRIES`. `DISPATCH_MESSAGE_ENTRIES` is still used where appropriate. Let me know if you think it should be switched elsewhere.

tt-aho committed:
#0: Split memcpy_to_device related functionality into new memcpy.hpp file

tt-aho committed:
#13655: Add SubDevice and SubDeviceManager implementations
Add support for splitting a device into multiple SubDevices, as well and maintaining different SubDeviceManager configurations, owned by device
Add basic tests to validate sub-device support
Update device apis to overload rather than take in optional sub_device ids
Make SubDeviceId, SubDeviceManagerId strong types
Refactor Device/SubDeviceManager state so that the default state is also encapsulated in a SubDeviceManager, and access the active SubDeviceManager through a pointer instead of map lookup

tt-aho committed:
#13655: Update allocator to support taking in a CoreRangeSet for banks, instead of assuming a rectangular grid
Update device allocator related apis to take in a sub-device parameter

patrickroberts commented on tt_metal/impl/device/device.cpp:
```diff
@@ -3405,6 +3553,113 @@ void Device::generate_device_headers(const std::string &path) const
+SubDeviceManagerId Device::create_sub_device_manager(tt::stl::Span<const SubDevice> sub_devices, DeviceAddr local_l1_size) {
+    static std::atomic<SubDeviceManagerId> sub_device_manager_id = 0;
+    TT_FATAL(!this->using_slow_dispatch(), "Using sub device managers is unsupported with slow dispatch");
+    auto [sub_device_manager, _] = this->sub_device_managers_.insert_or_assign(sub_device_manager_id, std::make_unique<detail::SubDeviceManager>(sub_devices, local_l1_size, this));
```
Seeing as this appears to be the only time the `SubDeviceManager` is ever constructed, is it necessary to keep the constructor overload that move-constructs a vector? I don't see it being used anywhere.

patrickroberts commented on tt_metal/tt_metal.cpp:
```diff
@@ -1137,37 +1150,38 @@ std::unique_ptr<GlobalSemaphore> CreateGlobalSemaphore(
-        config.buffer_layout,
-        config.shard_parameters,
-        std::nullopt);
+std::shared_ptr<Buffer> CreateBuffer(const InterleavedBufferConfig &config, std::optional<DeviceAddr> address, std::optional<uint32_t> sub_device_id) {
```
Please don't combine these overloads, I'd rather avoid the unnecessary `std::optional` branching here, where the logical was previously unconditional.

tt-aho replied:
Do you have a suggestion for avoiding this? I ended up combining them because the uint64_t address and The uint32_t sub_device_id are both integer type, so it was ambiguous to overload.

patrickroberts replied:
Make `sub_device_id` a strong type. `DeviceAddr` should have been a strong type too. Making them aliases for numeric types is what leads to completely avoidable issues like this.

tt-aho commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -2853,7 +3039,8 @@ void EnqueueReadBuffer(
     void* dst,
-    bool blocking) {
+    bool blocking,
+    tt::stl::Span<const uint32_t> sub_device_ids) {
```
Yes, will fix. I messed up when I was splitting my commits between the FD refactor and the sub-device implementation

tt-aho commented on tt_metal/impl/device/device.cpp:
```diff
@@ -3096,25 +3143,83 @@ uint32_t Device::get_noc_multicast_encoding(uint8_t noc_index, const CoreRange&
-uint32_t Device::num_banks(const BufferType &buffer_type) const {
-    this->check_allocator_is_initialized();
-    return allocator::num_banks(*this->allocator_, buffer_type);
+const std::unique_ptr<Allocator> &Device::get_initialized_allocator(std::optional<uint32_t> sub_device_id) const {
+    if (sub_device_id.has_value()) {
+        TT_FATAL(this->active_sub_device_manager_id_.has_value(), "Specifying sub-device id requires an active sub-device manager");
+        auto& sub_device_manager = this->sub_device_managers_.at(*this->active_sub_device_manager_id_);
+        TT_FATAL(*sub_device_id < sub_device_manager->num_sub_devices(), "Invalid sub-device id {}", *sub_device_id);
+        auto& allocator = sub_device_manager->sub_device_allocator(*sub_device_id);
+        if (!allocator) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+        return allocator;
+    } else {
+        if (!this->allocator_) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+        return this->allocator_;
+    }
+}
```
I'll replace it

tt-aho commented on tt_metal/impl/device/device.cpp:
```diff
@@ -3096,25 +3143,83 @@ uint32_t Device::get_noc_multicast_encoding(uint8_t noc_index, const CoreRange&
-void Device::check_allocator_is_initialized() const {
-    if (this->allocator_ == nullptr) {
-        TT_THROW("No memory allocator! Device has not been initialized, did you forget to call InitializeDevice?");
+void Device::check_allocator_is_initialized(std::optional<uint32_t> sub_device_id) const {
+    if (sub_device_id.has_value()) {
+        TT_FATAL(this->active_sub_device_manager_id_.has_value(), "Specifying sub-device id requires an active sub-device manager");
+        auto& sub_device_manager = this->sub_device_managers_.at(*this->active_sub_device_manager_id_);
+        TT_FATAL(*sub_device_id < sub_device_manager->num_sub_devices(), "Invalid sub-device id {}", *sub_device_id);
+        auto& allocator = sub_device_manager->sub_device_allocator(*sub_device_id);
+        if (!allocator) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
+    } else {
+        if (!this->allocator_) {
+            TT_THROW("No memory allocator! Allocator has not been initialized");
+        }
     }
 }
```
Yes, will separate

patrickroberts commented on tt_metal/impl/device/device.cpp:
```diff
@@ -2895,6 +2934,12 @@ bool Device::close() {
         }
+    }
+    this->clear_loaded_sub_device_manager();
+    for (auto sub_device_manager = this->sub_device_managers_.begin(); sub_device_manager != this->sub_device_managers_.end();) {
+        this->remove_sub_device_manager((sub_device_manager++)->first);
+    }
```
https://en.cppreference.com/w/cpp/container/unordered_map/erase:
> References and iterators to the erased elements are invalidated. **Other iterators and references are not invalidated.** (emphasis added)

tt-aho replied:
This does not appear to work.
Ex. In one of my tests I see the original loop remove ids 2, 1, 0.
Replacing this code with the suggestion I see it clear 2, then a random value causing it to error out.

tt-aho committed:
#13655: Refactor dispatching of go signal to not send noc txn data after go signal command
Instead, we populate a static array on dispatcher when we change sub-device configurations with all noc txn data, and read from it using an offset passed in the go signal command
Remove dynamic allocation of sub-devices/expected workers pairs, and pass them as separate spans
Fix cmd in sweep_pgm_dispatch

abhullar-tt commented on tt_metal/tt_metal.cpp:
```diff
@@ -1139,14 +1154,37 @@ std::unique_ptr<GlobalSemaphore> CreateGlobalSemaphore(
+        config.buffer_layout,
+        std::nullopt,
+        std::nullopt,
+        std::nullopt);
```
this can probably be cleaned up in a separate PR but could we add these extra params to the config? I think some are common between Interleaved and Sharded configs but having a bunch of nullopts and then sub device ids looks awkward. Could you create an issue for this

tt-aho replied:
Yes, I'll file an issue for this. The three nullopts are shard_parameters, bottom_up (this can probably go into the config), and sub_device_id. Do we want sub_device_id in the config as well, or does it make more sense as it is right now where host_api takes it as a separate argument like device addr?

abhullar-tt replied:
since it isn't really parameter of how to set up the buffer I think its fine for it to go outside the config
